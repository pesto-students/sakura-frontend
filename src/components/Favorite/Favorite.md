```js
<Favorite size={2} sizeUnit="rem" favoriteItemLength={1}  favoriteCbk = {(q) => { console.log(q)}}></Favorite>
```


