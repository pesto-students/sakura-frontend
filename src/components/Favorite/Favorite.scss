@import "../../sassStyles/globals";

.primary_favorite{
    --size: 2rem;
    --icon-size: 3rem;
    position: relative;
    display: inline-block;
    cursor: pointer;

    &_icon{
        display: inline-block;
        transition: all .2s;
        width: var(--icon-size);
        height: var(--icon-size);
        // color: rgba(var(--color-black--rgb), 1);
        &:hover{
            color:rgba($primary-color, 1);
        }

    }
    &_counter{
        display:inline-block;
        position: absolute;
        top: -30%;
        right: -30%;
        font-size: .8rem;
        width: var(--size);
        height: var(--size);
        line-height: var(--size);
        text-align: center;
        color: rgba($color-white, 1);
        background-color:rgba($primary-color, 1);
        border-radius: 50%;       
    }
}