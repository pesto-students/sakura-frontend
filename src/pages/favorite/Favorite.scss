@import "../../sassStyles/globals";

.primary_favorite{
    width: 100%;
    &_heading{
        margin: 2rem auto;
        font-size: 2rem;
        text-align: center;
        color:$secondary-color;
    }
    &_details{
        margin: 1rem 15vw;

        &_checkout{
            position: -webkit-sticky;
            position: sticky;
            top: 0;
            display: grid;
            width: 100%;
            display: grid;
            box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
            padding: 0.2rem 1rem;
            grid-template-columns: 1fr 1fr 1fr;
            grid-template-rows: repeat(6, min-content);
            border-radius: 4px;

            &_heading{
                grid-column: 1/4;
                font-size: 1.8rem;
                font-weight: 400;
            }
            &_item_amount_heading {
                grid-column: 1/3;
                font-size: 1.3rem;
            }

            &_item_amount_total {
                grid-column: 3/4;
                font-size: 1.3rem;
                justify-self: end;
                
            }

            &_shipping_amount_heading{
                grid-column: 1/3; 
                font-size: 1.3rem;
                

            }
            &_shipping_amount_total{
                grid-column: 3/4;
                font-size: 1.3rem;
                justify-self: end;
                
                
            }

            &_tax_amount_heading{
                grid-column: 1/3;
                font-size: 1.3rem;

            }
            &_tax_amount_total{
                grid-column: 3/4;
                font-size: 1.3rem;
                justify-self: end;
                
                
            }
            &_divider{
                align-self: center;
                grid-column: 1/4;
            }

            &_total_amount_heading{
                grid-column: 1/3;
                font-size: 1.3rem;
                align-self: start;
                padding-bottom: 1rem;
            }

            &_total_amount_total{
                grid-column: 3/4;
                font-size: 1.3rem;
                justify-self: end;
                
                font-weight: 400;
                align-self: start;
                padding-bottom: 1rem;
            }
            &_payment_button{
                grid-column: 1/4;
                padding-bottom: 2rem;
            }
        }

        &_empty{
        font-size: 1.5rem;
        text-align: center;
        color:$secondary-color;
        padding: 1rem;
        }
    }
    
}   