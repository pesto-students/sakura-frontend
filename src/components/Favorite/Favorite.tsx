import React from "react"
import { Heart } from 'react-bootstrap-icons';
import "./Favorite.scss"


/**
 * This component tells how many items are there in user cart
 */
export default class Favorite extends React.Component<FavoriteProps>{

    constructor(props: FavoriteProps) {
        super(props);
        this.state = {};
    }


    render () {

        const {size=2, sizeUnit="rem", favoriteItemCount: favoriteItemLength=0} = this.props;
        let iconSize: string = String(size*1.3) + sizeUnit
        let counterSize: string = String(size) + sizeUnit
        
        return (
            <div className="primary_favorite" data-testid="favorite" onClick={e=>this.props.favoriteCbk()}>
                <Heart className="primary_favorite_icon" style={{width: iconSize, height:iconSize}}/>
                <div className="primary_favorite_counter" style={{width:counterSize, height:counterSize, lineHeight: counterSize}}>{favoriteItemLength}</div>
            </div>
        )
    }

}

export type FavoriteProps = {
    /**
     * Size defines the size of component on different screen.
     */
     size?: number;
     /**
      * Define unit of size
      */
     sizeUnit?: String;
     /**
      * favorite item length tells the number of item in user favorite.
      */
    favoriteItemCount: number;
      /**
     * Emits observables on favorite click. This callback can be listened, using subscribe,
     * to get user input and call api from it. Further operators like throttle, map etc can be applied
     * on the emitted observable
     */
    favoriteCbk: () => void;
    
}