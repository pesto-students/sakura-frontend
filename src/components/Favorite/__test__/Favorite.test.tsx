import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { render, fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import renderer from "react-test-renderer"
import Favorite from '../Favorite'

// const server = setupServer(
//   rest.get('/search', (req, res, ctx) => {
//     return res(ctx.json({}))
//   })
// )

// beforeAll(() => server.listen())
// afterEach(() => server.resetHandlers())
// afterAll(() => server.close())

test('Should  call Cart callback on click', async () => {
  
  render(<Favorite size={2} sizeUnit="rem" favoriteItemCount={10} favoriteCbk={favoriteCallCbk}/>)
  let st = ""

  function favoriteCallCbk() {
        st = "favorite clicked! go to cart page"
        expect(st).toEqual("favorite clicked! go to cart page")
  }

  const favoriteClick = screen.getByTestId("favorite");
  fireEvent.click(favoriteClick);

})


test('Should render cartItemLegth', async () => {
  
  const favoriteProps = {favoriteSize: 50, favoriteItemLength: 20, favoriteCbk: function () {}}
  render(<Favorite size={2} sizeUnit="rem" favoriteItemCount={favoriteProps.favoriteItemLength} favoriteCbk={favoriteProps.favoriteCbk}/>)
  const favElement = screen.getByTestId("favorite");
  expect(favElement).toBeInTheDocument()
  expect(favElement).toHaveTextContent("20")

})

test("matches snapshots", ()=> {
  const favoriteProps = {favoriteSize: 50, favoriteItemLength: 20, favoriteCbk: function () {}}
  const tree = renderer.create(<Favorite size={2} sizeUnit="rem" favoriteItemCount={favoriteProps.favoriteItemLength} favoriteCbk={favoriteProps.favoriteCbk}/>).toJSON()
  expect(tree).toMatchSnapshot()
})