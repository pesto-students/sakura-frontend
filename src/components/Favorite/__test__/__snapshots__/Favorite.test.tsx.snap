// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots 1`] = `
<div
  className="primary_favorite"
  data-testid="favorite"
  onClick={[Function]}
>
  <svg
    className="primary_favorite_icon"
    fill="currentColor"
    height="1em"
    style={
      Object {
        "height": "2.6rem",
        "width": "2.6rem",
      }
    }
    viewBox="0 0 16 16"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 2.748l-.717-.737C5.6.281 2.514.878 1.4 3.053c-.523 1.023-.641 2.5.314 4.385.92 1.815 2.834 3.989 6.286 6.357 3.452-2.368 5.365-4.542 6.286-6.357.955-1.886.838-3.362.314-4.385C13.486.878 10.4.28 8.717 2.01L8 2.748zM8 15C-7.333 4.868 3.279-3.04 7.824 1.143c.06.055.119.112.176.171a3.12 3.12 0 0 1 .176-.17C12.72-3.042 23.333 4.867 8 15z"
    />
  </svg>
  <div
    className="primary_favorite_counter"
    style={
      Object {
        "height": "2rem",
        "lineHeight": "2rem",
        "width": "2rem",
      }
    }
  >
    20
  </div>
</div>
`;
